src/Chapter3.hs

```diff
@@ -373,6 +383,26 @@ after the fight. The battle has the following possible outcomes:
    doesn't earn any money and keeps what they had before.
 
 -}
+data Knight_ = Knight_
+  { knightHealth_ :: Int
+  , knightAttack_ :: Int
+  , knightGold_ :: Int
+  } deriving (Show)
+
+data Monster_ = Monster_
+  { monsterHealth_ :: Int
+  , monsterAttack_ :: Int
+  , monsterGold_ :: Int
+  } deriving (Show)
+
+fight_ :: Monster_ -> Knight_ -> Int
+fight_ (Monster_ mHealth mAttack mGold) (Knight_ kHealth kAttack kGold)
+  | mHealth <= 0 = kGold + mGold
+  | kHealth <= 0 = -1
+  | otherwise    = fight_ monster knight
+  where
+    monster = Monster_ (mHealth - kAttack) mAttack mGold
+    knight  = Knight_ (kHealth - mAttack) kAttack kGold
```

Comment:
Nice! Here you can use record update syntax:

```haskell
    knightNew  = knight { kHealth = kHealth - mAttack}
```
For that in the function definition you can name the parameter of knight as `knight`.
One way to do that is to use `@` to name the whole argument:

```haskell
fight_ monster@(Monster_ mHealth mAttack mGold) knight@(Knight_ kHealth kAttack kGold)
```

Additionally, Haskell has an extension (I saw you already tried to use other extensions in this chapter, so I assume that you are familiar with the concept) – `RecordWildCards`. This extension allows to use named fields of the record, as values, without need to manually pattern much as you do:

```haskell
{-# LANGUAGE RecordWildCards #-}

fight_ Monster_{..} Knight_{..}
  | mHealth_ <= 0 = kGold_ + mGold_
...
```

You can read more about the extension on this article: https://kodimensional.dev/recordwildcards

Reply:
Cool, like some black magic!